src/ai/backend/manager/actions/monitors/smtp.py

```diff
@@ -0,0 +1,121 @@
+import asyncio
+import enum
+from dataclasses import dataclass
+from datetime import datetime
+from email.mime.text import MIMEText
+from typing import Final, override
+
+import aiosmtplib
+
+from ai.backend.manager.models.audit_log import OperationStatus
+
+from ...actions.action import BaseAction, ProcessResult
+from .monitor import ActionMonitor
+
+UNKNOWN_ENTITY_ID: Final[str] = "(unknown)"
+
+
+class SMTPTriggerPolicy(enum.Flag):
+    PRE_ACTION = enum.auto()
+    POST_ACTION = enum.auto()
+    ON_ERROR = enum.auto()
+
+
+@dataclass
+class SMTPReporterConfig:
+    host: str
+    port: int
+    username: str
+    password: str
+    sender: str
+    recipients: list[str]
+    use_tls: bool
+    trigger_policy: SMTPTriggerPolicy
+
+
+class SMTPReporter(ActionMonitor):
+    def __init__(self, config: SMTPReporterConfig, concurrency_limit: int = 5) -> None:
+        self._config = config
+        self._semaphore = asyncio.Semaphore(concurrency_limit)
+        self._smtp = aiosmtplib.SMTP(
+            hostname=self._config.host, port=self._config.port, use_tls=self._config.use_tls
+        )
+        self._smtp_lock = asyncio.Lock()
+
+    def _make_action_type(self, action: BaseAction) -> str:
+        return f"{action.entity_type()}:{action.operation_type()}"
+
+    def _make_subject(self, action: BaseAction) -> str:
+        return f"Backend.AI SMTP Log Alert ({self._make_action_type(action)})"
+
+    async def _send_email(self, subject: str, email_body: str) -> None:
+        async with self._semaphore:
+            email_body += "\nThis email is sent from Backend.AI SMTP Reporter"
+
+            message = MIMEText(email_body, "plain", "utf-8")
+            message["Subject"] = subject
+            message["From"] = self._config.sender
+            message["To"] = ",".join(self._config.recipients)
+
+            try:
+                await self.connect_smtp()
+                await self._smtp.send_message(
+                    message, sender=self._config.sender, recipients=self._config.recipients
+                )
+            except Exception as e:
+                print(f"Failed to send email: {e}")
```

Comment:
Personally, I think it would be better to use `asyncio.Queue` internally rather than code that uses semaphores.